\documentclass[a4paper]{article}

%\VignetteEngine{knitr::knitr}

\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{rotating}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage[a4paper, hdivide={2cm,,2cm}, vdivide={2cm,,2cm}]{geometry}
\usepackage{parskip}
\usepackage{natbib}
\usepackage{tabularx}
\usepackage{fancyvrb}
\usepackage{multicol}
\usepackage[hidelinks]{hyperref}


\title{A short introduction to the qpcrnlme package}
\date{\today}
\author{Daniel Gerhard}


\begin{document}


%\VignetteIndexEntry{Using qpcrnlme} \\
%\VignettePackage{qpcrnlme} \\


\maketitle

\section{Data Examples}

\subsection{KOTWTcurves}

For this experiment an insertion has been introduced into the genome of different rice lines. It is assumed that because of this insertion, the expression of the gene 10g30610 is changed. By using the quantitative RT-PCR methodology the changes in gene expression of the target gene is analysed between two different insertions and the wild type. \texttt{Content}: ID of biological replicates; \texttt{Target}: gene of interest and EF (elongation factor) gene of control; \texttt{Samples}: rice lines with an insertion (KO, T) and wild type (W).

<<loaddataKOTWTcurves, message=FALSE, fig.width=12, fig.height=6>>=
library(qpcrnlme)
data(KOTWTcurves)
str(KOTWTcurves)
KOTWTcurves$Sample <- factor(KOTWTcurves$Sample, levels=c("W", "T", "KO"))
ggplot(KOTWTcurves, aes(x=Cycle, y=Fluorescence, colour=Sample, group=Well)) +
  geom_line()
@

\subsection{pt6c}

In this experiment rice ({\it Oryza sativa} L.) plants were starved five days for phosphate resulting in an upregulation of the phosphate transporter PT6 to increase the uptake of the limited nutrient. Resupply of phosphate decreased the expression of PT6 after 0, 2, 4, 6, and 8 hours (variable \texttt{Time}). At each time point the expression of PT6 and the reference gene eEf (variable \texttt{Target}) were observed for three biological replicates (variable \texttt{Content}), which are thought of as representatives of the underlying biological system. For each biological replicate and gene the fluorescence intensity (variable \texttt{Fluorescence}) was measured over 40 PCR cycles (variable \texttt{Cycle}) for three technical replicates (variable \texttt{Well}).

<<loaddatapt6c, message=FALSE, fig.width=12, fig.height=6>>=
data(pt6c)
str(pt6c)
ggplot(pt6c, aes(x=Cycle, y=Fluorescence, colour=Time, group=Well)) +
  geom_line()
@

You may want to skip the model description and directly continue with the analysis of the data example in Section \ref{TreatmentComparisons}.

\section{Modeling RT-PCR Data}

We consider fluorescence intensities $\{ y_{ijk}\}$ obtained from a hierarchical design using a number of biological replicates ($i=1,\ldots,I$), which are again divided into a number of technical replicates ($j=1,\dots,J$), each with PCR cycle numbers $c_{k}$ ($k=1,\ldots,K$).

\subsection{Modeling the Fluorescence Curve}

At the first stage we specify the nonlinear relationship for each individual fluorescence intensity:
\begin{equation} 
y_{ijk} = f\left(c_{k}, \boldsymbol{\beta}_{ij} + \boldsymbol{u}_{ij} \right) + \epsilon_{ijk}, 
\end{equation}
assuming a specific nonlinear model function $f$ depending on the cycle number $c_{k}$ and $R$-dimensional vectors of fixed- and random-effects contributions $\boldsymbol{\beta}_{ij}$ and $\boldsymbol{u}_{ij}$, respectively. The error terms follow a normal distribution $N\left( 0, \sigma^{2} \right)$. 

Until now, only the five-parameter log-logistic model is implemented in qpcrnlme:
\begin{align} \label{LL.5} 
f \left(c_k, \boldsymbol{\beta}_{ij} \right) = & f \{ c_k, (\beta_{ij}^{(1)}, \ldots, \beta_{ij}^{(5)}) \}\\
= &\beta_{ij}^{(2)} + \frac{\beta_{ij}^{(3)}-\beta_{ij}^{(2)}}{(1+\exp[\beta_{ij}^{(1)}\{\log(c_{j})-\log(\beta_{ij}^{(4)})\} ])^{\beta_{ij}^{(5)}}} \nonumber
\end{align}
where $\beta_{1ij}$ characterize the steepness in the s-shaped curve, $\beta_{2ij}$ and $\beta_{3ij}$ correspond to the lower and upper asymptotes, $\beta_{4ij}$ denotes the approximate location of the inflection point, and $\beta_{5ij}$ is an asymmetry parameter where positive values above or below 1 correspond to differences in curvature close to the lower and upper asymptotes. 

This function can be evaluated by calling \texttt{logistic5()}. For example, the effect of the asymmetry parameter on the logistic function can be illustrated.
<<LL5, fig.width=12, fig.height=6>>=
cycle <- seq(1, 40, by=1)
f05 <- llogistic5(cycle, b=-10, c=0, d=750, e=20, f=0.5)
f1 <- llogistic5(cycle, b=-10, c=0, d=750, e=20, f=1)
f2 <- llogistic5(cycle, b=-10, c=0, d=750, e=20, f=2)
asdat <- data.frame(cycle, response=c(f05, f1, f2), asymmetry=factor(rep(c(0.5, 1, 2), each=length(cycle))))
ggplot(asdat, aes(x=cycle, y=response, colour=asymmetry)) + geom_line()
@

\subsection{Hierarchical Structure of Biological and Technical Replicates}

At the second stage we specify the decomposition of the model parameters at the level of the technical replicates. For the $r$th parameter in the vector $\boldsymbol{\beta}_{ij}$ the fixed- and random effects contributions are:
\begin{align*}
\boldsymbol{\beta}^{(r)}_{ij} = &\left(\boldsymbol{X}^{(r)}_{T} \right)_j \otimes \left(\boldsymbol{X}^{(r)}_{B} \right)_i \boldsymbol{\beta},\\
%&\left( \boldsymbol{X}^{(r)}_{B} \right)_i \boldsymbol{\beta^{(r)}_B} + \left(\boldsymbol{X}^{(r)}_{T}\right)_j \boldsymbol{\beta^{(r)}_T}\\
%& + \left(\boldsymbol{Z}^{(r)}_{B}\right)_i \boldsymbol{u}^{(r)}_{B} + \left(\boldsymbol{Z}^{(r)}_{T}\right)_j\boldsymbol{u}^{(r)}_{T},
\boldsymbol{u}^{(r)}_{ij} = &\left(\boldsymbol{Z}^{(r)}_{B}\right)_i \boldsymbol{u}^{(r)}_{B} + \left(\boldsymbol{Z}^{(r)}_{T}\right)_j \boldsymbol{u}^{(r)}_{Ti},
\end{align*}
where $\boldsymbol{X}^{(r)}_{B}$ $(I \times p_1)$ and $\boldsymbol{X}^{(r)}_{T}$ $(J \times p_2)$ denote the design matrices of the fixed-effects structures at the level of the biological and technical replicates, respectively (the subscript refers to specific rows in the matrices), and $\boldsymbol{\beta}$ denotes the $p_1p_2$-dimensional fixed-effects parameter. Any kind of covariate information available at the level of the biological replicates could be included in the model. For our data example $\boldsymbol{X}^{(r)}_{B}=\boldsymbol{X}_{B}$ specifies a time trend that is assumed to be present in all $r$ parameters, whereas $\boldsymbol{X}^{(r)}_{T} = \boldsymbol{X}_{T}$ groups technical replicates corresponding to the same gene into clusters (two clusters within each biological replicate as we consider two genes). 

Similarly, $\boldsymbol{Z}^{(r)}_{B}$ ($I \times q_1$) and $\boldsymbol{Z}^{(r)}_{T}$ ($J \times q_2$) are the design matrices of the random effects associated with the biological and technical replicates within biological replicates, respectively. The random effects are assumed to follow normal distributions:
\begin{align*}
\boldsymbol{u}^{(r)}_B & \sim N(\boldsymbol{0}, \boldsymbol{\Psi}^{(r)}_{B}),\\
\boldsymbol{u}^{(r)}_{T} = (\boldsymbol{u}^{(r)}_{T1}, \ldots, \boldsymbol{u}^{(r)}_{TI}) & \sim N (0, \boldsymbol{\Psi}^{(r)}_{T}). 
\end{align*}
In principle $\boldsymbol{\Psi}^{(r)}_{B}$ and $\boldsymbol{\Psi}^{(r)}_{T}$ may be unstructured variance-covariance matrices, but we restrict ourselves to diagonal matrices that correspond to uncorrelated random effects, apart from letting $\boldsymbol{\Psi}^{(r)}_{B}$ be a block diagonal matrix with diagonal entries that are themselves diagonal matrices $\boldsymbol{\Psi}^{(r)}_{B1}, \hdots, \boldsymbol{\Psi}^{(r)}_{Bp_2}$ allowing different variance components for different groups \citep[pp.~122--124]{Davidian1995}. 


\subsection{Implementation}

To fit a hierarchical nonlinear model to the RT-PCR data, the qpcrnlme package makes use of the function \texttt{nlme} in the package \texttt{nlme}. For a fixed layout of the RT-PCR experiment, most of the model parameters are pre-specified, like a common upper asymptote for each gene and treatment, and the random effect structure for biological and technical replicates with different variance components for each gene.
Starting values are found automatically by estimating several nonlinear models for each random effect level with the package drc; hence, no additional starting value has to be provided. 


\section{Marginal Cycle Number Estimation}

It is established practice to evaluate real-time PCR data by means of the threshold cycle summary measure instead of directly interpreting the parameters in $f$. The threshold cycle is defined as the cycle number where the mean fluorescence level reaches a certain cutoff intensity $t$. Determining the threshold is an inverse regression problem that in case of a nonlinear regression model for a single replicate has the solution $c(t) = f^{-1}(t)$. For a nonlinear mixed model, the threshold cycle is obtained by solving the equation $E(f(c, \boldsymbol{\beta}_{0ij} + \boldsymbol{u}_0)) = t$ in $c$ for some specific fixed-effects parameter configuration denoted $\boldsymbol{\beta}_{0ij}$. Solving the equation requires repeated evaluation of the integral by integrating out the random effects $\boldsymbol{u}_0$.
The qpcrnlme package uses Gauss-Hermite quadrature to approximate these integrals with help of the package statmod.

The result of a RT-PCR data analysis with the qpcrnlme function is mainly a vector with marginal $c(t)$ estimates and the corresponding variance-covariances. Based on these estimates, specific linear combinations of the $c(t)$ values can be constructed. $\Delta\Delta c(t)$ values to compare treatment levels and genes can be calculated by function ddctcomp. 
For these derived parameters, hypotheses tests and confidence intervals are provided.


\section{Treatment Comparisons}
\label{TreatmentComparisons}

\subsection{Pairwise-Comparisons to a Control}

For a first evaluation, the two different insertion treatment groups in the KOTWTcurves dataset are compared to the wildtype group (W). The sample variable is a factor for this example, hence the design matrices for the fixed effects are structured in a way that for each gene-treatment level combination a single $c(t)$ value is estimated. 

The model parameterization is defined by assigning the appropriate column names of the data.frame to the corresponding variable input argument, that is, a fluorescence response, a vector with cycle numbers, identifiers of treatment and gene factors, and factors specifying the structure of biological and technical replicates. 

Some optional arguments can be changed:
\begin{itemize}
\item With the cutoff argument a $c(t)$ cutoff $t$ can be specified. This value should lie within the range of the lower and upper asymptote. 
\item The nGQ defines the number of nodes and weights for the Gauss-Hermite approximation.
\end{itemize}

The marginal $c(t)$ estimates for each time group per gene are calculated by
<<comparisonctest, warning=FALSE>>=
ctest <- qpcr_nlme(response="Fluorescence", cycle="Cycle", gene="Target", treatment="Sample", brep="Content", well="Well", data=KOTWTcurves, cutoff=100, nGQ=1, verbose=FALSE)
print(ctest)
@


$\Delta\Delta c(t)$ comparisons to the control can be performed by the function ddctcomp. Besides the definition of a control gene and control treatment levels, following arguments can be modified:
\begin{itemize}
\item The conf.level argument specifies a confidence level of a $\Delta\Delta c(t)$ confidence interval. 
\item If the \verb ratio_ddct  argument has the value TRUE, ratios of $c(t)$ estimates are computed to define the $\Delta\Delta c(t)$, otherwise the difference of $c(t)$ values are calculated. 
\item If adjusted is TRUE, the family-wise error rate is controlled for all teatment comparisons. A single-step procedure is used, similar to the implementation in package multcomp, to obtain adjusted p-values and simultaneous confidence intervals. If adjusted equals FALSE, the per comparison error rate is used, assigning a separate type-I-error rate to each comparison.
\end{itemize}

<<comparison3control, warning=FALSE>>=
ddctcomp(ctest, control_treatment="W", control_gene="EF", conf.level=0.95, ratio_ddct=TRUE, adjusted=TRUE)
@

\subsection{All Pairwise-Comparisons}

If no control treatment or control gene can be specified, the control lables can be set to NULL. In this case, all pairwise-comparison of treatment and/or gene groups are performed. With only two genes there will be no difference for the genewise comparisons to the comparisons to a control.

<<allpaircomparisons, warning=FALSE>>=
ddctcomp(ctest, control_treatment=NULL, control_gene=NULL, conf.level=0.95, ratio_ddct=TRUE, adjusted=TRUE)
@


\subsection{General Linear Contrasts}

To provide the ability to define more general linear contrasts, as described for RT-PCR analysis in \citet[p. 151]{Steibel2009}, the R package multcomp can be used directly with the nlmect class. For example, the KO genotype can be compared with the average of the T and W group, for each single gene, and for the interaction contrast, comparing the differential expression of the treatments based on the comparison of gene 10g30610 with EF. Thus, for each $c(t)$ estimate in the \verb ctest  object, a contrast coefficient is assigned to compose the desired linear combination of $c(t)$ parameters. Simultaneous confidence intervals and adjusted p-values are given for each defined contrast. 

<<userdefinedcontrasts, warning=FALSE, message=FALSE, tidy=FALSE>>=
library(multcomp)
K <- rbind("KO vs. ave(T, W) | 10g"=c(1, -0.5, -0.5, 0, 0, 0),
           "KO vs. ave(T, W) | EF"=c(0, 0, 0, 1, -0.5, -0.5),
           "KO vs. ave(T, W) | 10g vs EF"=c(1, -0.5, -0.5, -1, 0.5, 0.5))
gc <- glht(ctest, K)
summary(gc)
confint(gc)
@

If instead of differences, the ratios of linear combinations of parameters are of interest, the gsci.ratio function in the package mratios can be applied, making use of Fiellers theorem to compute the simultaneous confidence intervals.


\section{More general modeling options}

If a more specific specification of the designmatrix of the fixed effects is needed, a formula interface to model the treatment effects per gene is available in the function \verb qpcr_nlme_formula . Until now this interface is only restricted to apply the same design matrices for each of the nonlinear model parameters.

The pt6c dataset is used to illustrate the function \verb qpcr_nlme_formula , which enables us to describe the change in differential expression over time by a regression model.

As an example, a cubic polynomial function is used to describe the differential gene-expression in time separately for each gene and each nonlinear model parameter except the upper asymptote.

<<poly4, warning=FALSE>>=
poly4 <- qpcr_nlme_formula(response="Fluorescence", cycle="Cycle", gene="Target", trtformula=~ poly(Time, 3, raw=TRUE), brep="Content", well="Well", data=pt6c, newdata=data.frame(Time=seq(1,8, length=5)), cutoff=100, nGQ=1, verbose=FALSE)
print(poly4)
@



\section{Structure of a ddct Object}

A ddct object contains all results of each modeling step. The content of this class of objects is shown for the comp2control results.

The output of the nonlinear mixed model is found in the nlme slot. Here, access to the estimated variance components, fixed- and random effects, fitted values, etc. is available. 
<<ddctnlme>>=
print(ctest$nlme, correlation=FALSE)
@

The marginal $c(t)$ estimates can be obtained from the slot ct, the corresponding variance-covariance matrix is available in the vcov slot.
<<ddctctest>>=
print(ctest$ct)
print(ctest$vcov)
@





\bibliographystyle{plainnat}  
\begin{thebibliography}{10}
\bibitem[\protect\citeauthoryear{Bates and Watts}{1988}]{Bates1988}
Bates, D. M. and Watts, D. G. (1988). Nonlinear regression analysis and its applications, Wiley, New York.
\bibitem[\protect\citeauthoryear{Davidian and Giltinan}{1995}]{Davidian1995}
Davidian, M. and Giltinan, D. M. (1995). Nonlinear Models for Repeated Measurement Data, Chapman and Hall, London.
\bibitem[\protect\citeauthoryear{Ritz}{2010}]{Ritz2010}
Ritz, C. (2010). Towards a unified approach to dose-response modeling in ecotoxicology. 
{\it Environmental Toxicology \& Chemistry} {\bf 29,} 220--229.
\bibitem[\protect\citeauthoryear{Spiess et al.}{2008}]{Spiess2008}
Spiess A. N., Feig C., Ritz C. (2008). Highly accurate sigmoidal fitting of
  real-time PCR data by introducing a parameter for asymmetry. {\it BMC Bioinformatics} {\bf 9,} 221.
\bibitem[\protect\citeauthoryear{Steibel et al.}{2009}]{Steibel2009}
Steibel J. P., Poletto R., Coussens P. M., Rosa G. J. M. (2009). A powerful and flexible
  linear mixed model framework for the analysis of relative quantification
  RT-PCR data. {\it Genomics} {\bf 94,} 146--152.  
\end{thebibliography}




\end{document}